src/Platform.Handlers/src/Xamarin.Platform.Handlers/Hosting/Internal/ServiceFactoryAdapter.cs

```diff
@@ -0,0 +1,51 @@
+﻿using System;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Hosting;
+
+namespace Xamarin.Platform.Hosting.Internal
+{
+	internal class ServiceFactoryAdapter<TContainerBuilder> : IServiceFactoryAdapter
+	{
+#pragma warning disable CS8714 // The type cannot be used as type parameter in the generic type or method. Nullability of type argument doesn't match 'notnull' constraint.
+		private IServiceProviderFactory<TContainerBuilder>? _serviceProviderFactory;
+		private readonly Func<HostBuilderContext>? _contextResolver;
+		private readonly Func<HostBuilderContext, IServiceProviderFactory<TContainerBuilder>>? _factoryResolver;
+
+		public ServiceFactoryAdapter(IServiceProviderFactory<TContainerBuilder> serviceProviderFactory)
+		{
+			_serviceProviderFactory = serviceProviderFactory ?? throw new ArgumentNullException(nameof(serviceProviderFactory));
+		}
+
+		public ServiceFactoryAdapter(Func<HostBuilderContext> contextResolver, Func<HostBuilderContext, IServiceProviderFactory<TContainerBuilder>> factoryResolver)
+		{
+			_contextResolver = contextResolver ?? throw new ArgumentNullException(nameof(contextResolver));
+			_factoryResolver = factoryResolver ?? throw new ArgumentNullException(nameof(factoryResolver));
+		}
+
+		public object CreateBuilder(IServiceCollection services)
+		{
+			if (_serviceProviderFactory == null)
+			{
+				_serviceProviderFactory = _factoryResolver!(_contextResolver!());
```

Comment:
Just a note on what I did here to remove the warnings. I added a `!` or null-forgiving operator to pinky swear we know what we are doing.